### ISSUE TITLE
DO CONCURRENT might be broken

### ISSUE THREAD

**Issue by certik** [j3-fortran/fortran_proposals#62]
@klausler reported in https://github.com/j3-fortran/fortran_proposals/issues/60#issuecomment-548613134:

> `DO CONCURRENT` is fundamentally broken. It only guarantees to the compiler that the iterations of the loop can be run in any serial order. Its default localization rule (any variable read in an iteration will see the most recent value written in the same iteration, if any) prevents straightforward parallel execution.

Let's discuss that here. @klausler, can you work with @gklimowicz to fix that? Gary has some proposals regarding "do concurrent". Or there is no way to fix this issue.

**Comment by klausler** [comment]
See https://j3-fortran.org/doc/year/19/19-134.txt .

**Comment by klausler** [comment]
I received no official response.

**Comment by certik** [comment]
@klausler thank you. As a member of the committee, I apologize. This is unacceptable to me and I am trying to convince the committee that we need to consider every proposal that gets officially submitted (even if for just 5 to 10 minutes). In fact I was at the February 2019 meeting, but I don't recall what happened to your paper, as that was my first meeting and I was just trying to figure out how the committee works. Now when we have this GitHub repository, I plan to track every technical comment the committee makes in issues.

If it makes you feel any better, the committee didn't consider my proposal either (in #1). I know it happened to others too. I feel it's very inefficient, because had the committee provided feedback to you, you could have submitted a better paper for the October 2019 meeting, and so on, and we could have had this feature in a much more "ready" shape.

@sblionel, here is an example of a proposal that I suggest the committee spends 5 to 10 minutes at plenary to discuss and then I volunteer to summarize the feedback in this issue here.

**Comment by klausler** [comment]
My problem, as an implementor, is that Fortran users expect from its name that `DO CONCURRENT` is sufficient to guarantee to the compiler that loop iterations can be safely executed in parallel; but the standard only suffices to guarantee that the iterations can be executed in arbitrary serial order.  What are my options?  (1) just parallelize it anyway; (2) provide an option or directive or extension to force parallelization in the face of unresolvable localizations; (3) emit a warning about unresolvable localizations and recommend the use of OpenMP directives instead?  People expect `DO CONCURRENT` to just mean "do concurrently", and it needs to be fixed or replaced so that it does.

**Comment by certik** [comment]
@klausler I agree. Let's discuss it further. Your proposal has an example:
```fortran
SUBROUTINE FOO(N, A, B, T, K, L)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: N, K(N), L(N)
  REAL, INTENT(IN) :: A(N)
  REAL, INTENT(OUT) :: B(N)
  REAL, INTENT(INOUT) :: T(N)
  INTEGER :: J
  DO CONCURRENT (J=1:N)
    T(K(J)) = A(J)
    B(J) = T(L(J))
  END DO
END SUBROUTINE FOO
```
How would this be written if your proposal is accepted? Using the `LOCAL` specifier.

**Comment by klausler** [comment]
I should correct this case; I intend `T()` to be dead after the loop.

The validity of this loop depends on the values of `K()` and `L()`.  Today, this loop is valid if the elements of `K()` are pairwise distinct from each other and from the noncorresponding elements of `L()` (obviously), and also in the case that `K(J)==L(J)` for all `J`.  Even the extreme case of `K(J)==L(J)==CONSTANT` is fine.

I would restrict `DO CONCURRENT` to be valid only under the first condition.  There's wouldn't be recourse to a `LOCAL` clause.  If you want the compiler to automatically localize a temporary variable, it would have to be an obvious whole scalar variable, not an array element.

**Comment by certik** [comment]
@klausler let's write down the actual examples, it's will be much easier for me (and I am sure others) to follow the arguments.

1)

```fortran
K = [1, 2, 3]
L = [4, 5, 6]
```

2) 

```fortran
K = [1, 2, 3]
L = [1, 2, 3]
```


3) 

```fortran
K = [1, 1, 1]
L = [1, 1, 1]
```

Currently 1., 2., and 3. are allowed. You are proposing for 1. to be allowed, but 2. and 3. to be forbidden. Given that this depends on the values of the arrays `K ` and `L`, which are only known at runtime, how would the compiler enforce it?

Isn't the idea of `do concurrent` that the user is responsible for ensuring that the loops are independent, which in many cases (such as this one) is impossible to do for the compiler, but quite possible for the user, since he knows his problem?

**Comment by klausler** [comment]
> @klausler let's write down the actual examples, it's will be much easier for me (and I am sure others) to follow the arguments.
> 
> 
> ```fortran-modern
> K = [1, 2, 3]
> L = [4, 5, 6]
> ```
> 
> 
> ```fortran-modern
> K = [1, 2, 3]
> L = [1, 2, 3]
> ```
> 
> 
> ```fortran-modern
> K = [1, 1, 1]
> L = [1, 1, 1]
> ```
> 
> Currently 1., 2., and 3. are allowed. You are proposing for 1. to be allowed, but 2. and 3. to be forbidden. Given that this depends on the values of the arrays `K ` and `L`, which are only known at runtime, how would the compiler enforce it?
> 
> Isn't the idea of `do concurrent` that the user is responsible for ensuring that the loops are independent, which in many cases (such as this one) is impossible to do for the compiler, but quite possible for the user, since he knows his problem?

No, I'm proposing only that your third case be disallowed.  And I don't know what the original idea behind `DO CONCURRENT` was, but I think that the users and implementors that I deal with would agree with your characterization of the obligations that it places on the compiler and programmer (basically the same as an OpenMP parallel loop).  The bug in `DO CONCURRENT` is that the compiler has to allow for your third case and make it work, either by running the code serially or detecting it and handling it dynamically.

**Comment by certik** [comment]
I corrected my comment above to match what you are proposing.

Regarding your last edit `K==L==[1,1,2]`, let's call this a case 4. Isn't case 4 the same as case 3 that should be disallowed? I am still a bit confused.

**Comment by klausler** [comment]
When all of the elements of `K()` are pairwise distinct, and distinct from all of the noncorresponding elements of `L()`, the loop is safe to parallelize.

*Otherwise*, and this is the problem, `DO CONCURRENT` also permits elements of `K()` to match each other, *so long as `K(J)==L(J)`*.  This is the problem.  (The case of *all* elements of `K()` and `L()` being the same value is a degenerate case of this second situation.)

So `K==[1,2,3],L==[4.5,6]` is fine; so are `K==[1,2,3],L==[1,5,6]` and `K==[1,2,3],L==[1,2,3]`.

But `K=[1,1,2],L=[1,1,2]` should not be; nor should `K==[1,1,1],L=[1,1,1]`, and these are the cases that I think should be invalidated.

**Comment by certik** [comment]
Thank you. I think it is clear now what cases are allowed and which ones are not.

Let's get some feedback from other members of the committee.

@sblionel, do you know the original idea behind `do concurrent` was and if what we are proposing here is aligned with it? See @klausler's previous comment: https://github.com/j3-fortran/fortran_proposals/issues/62#issuecomment-549991917, and the rest of the discussion here.

If other members of the committee agree that this should be fixed, then the next step is to update the proposal. I am happy to help.

**Comment by sblionel** [comment]
This discussion puzzles me somewhat, as I don't agree with some of the assertions made. However, this is not my area of expertise and I would prefer to see opinions of committee members more versed in parallelism, such as Bill Long of Cray.

DO CONCURRENT was designed as a replacement for F95's FORALL as it was determined that FORALL's semantics made parallelism very difficult. The whole idea of DO CONCURRENT is that the iterations in any order and to any degree of parallelism, as the user promises there are no cross-iteration dependencies. There are already Fortran implementations that successfully parallelize DO CONCURRENT (Intel and probably Cray), so I don't really understand what the problem is. DO CONCURRENT was mainly modeled on OpenMP PARALLEL DO, especially with the F18 additions of locality clauses.

I see that Peter's paper got "deferred" at the February 2019 meeting and not taken up again. I'm not a good person to discuss this with.

**Comment by klausler** [comment]
I suspect that the language in what is now 11.1.7.5 para. 4 (first bullet) was intended to handle cases like:

```
DO CONCURRENT (J=1:N)
  TMP = ...
  ... = TMP
END DO
```

by requiring the processor to automatically localize the obvious temporary `TMP`.  Unfortunately, the language is vague enough to also cover the cases that I've mentioned above where the "variable" in question can't be identified at compilation time.  It doesn't make a difference if the iterations of the loop are executed in some arbitrary serial order, but it does matter when parallelizing (or failing to parallelize because of the requirement in 11.1.7.5 para. 4, first bullet).

**Comment by klausler** [comment]
> This discussion puzzles me somewhat, as I don't agree with some of the assertions made. However, this is not my area of expertise and I would prefer to see opinions of committee members more versed in parallelism, such as Bill Long of Cray.
> 
> DO CONCURRENT was designed as a replacement for F95's FORALL as it was determined that FORALL's semantics made parallelism very difficult. The whole idea of DO CONCURRENT is that the iterations in any order and to any degree of parallelism, as the user promises there are no cross-iteration dependencies. There are already Fortran implementations that successfully parallelize DO CONCURRENT (Intel and probably Cray), so I don't really understand what the problem is. 

I will try to make this as clear as I possibly can, using the example that I presented earlier.

```
subroutine foo(N, A, B, T, K, L)
  implicit none
  integer, intent(in) :: N, K(N), L(N)
  real, intent(in) :: A(N)
  real, intent(out) :: B(N)
  real, intent(inout) :: T(N)
  integer :: J
  do concurrent (J=1:N)
    ! During execution, K(J) and L(J) are both always 1.  So the store
    ! and load to/from T() always affect T(1) in each iteration.  Since
    ! T(1) is defined in each iteration before it is referenced, this
    ! program conforms with F2008 and F2018.
    T(K(J)) = A(J)
    B(J) = T(L(J)) ! must be A(J) whenever K(J)==L(J)
  end do
end subroutine foo
```

This subroutine complies with all of the constraints and "shalls" in the Fortran 2018 standard that pertain to DO CONCURRENT.  But it cannot be executed in parallel and produce correct results.

This is because the DO CONCURRENT construct, despite its name, imposes restrictions on the program that are sufficient to guarantee that the iterations of the loop may be run *in any sequential order*.  The restrictions necessary to guarantee safe execution in arbitrary sequential order are not sufficient to guarantee safe execution in parallel.

When ifort parallelizes this loop (`ifort -parallel`), it will *fail* during execution when placed in a test harness.  (Alternatively, when this subroutine appears in the same source file as the test harness, ifort will expand it inline and then refuse to parallelize it due to a correctly diagnosed data dependence.)

> DO CONCURRENT was mainly modeled on OpenMP PARALLEL DO, especially with the F18 additions of locality clauses.

Except that OpenMP PARALLEL DO imposes stricter restrictions on the loop, so that it can be safely executed in parallel.  DO CONCURRENT's restrictions are only strong enough to ensure safe serial execution in arbitrary order of iterations.  Perhaps it was believed that these weaker restrictions would be easier to comply with, and that a sufficiently smart compiler could apply automatic localization and then safely parallelize.  This turns out _not to be the case_, since a compiler can apply automatic localization only to variables that can be identified at compilation time.
 
> I see that Peter's paper got "deferred" at the February 2019 meeting and not taken up again. I'm not a good person to discuss this with.

Why not?

**Comment by sblionel** [comment]
> Why not?

Because I don't head J3, nor even the HPC subgroup under whose purview this would fall.

As I wrote above, I do not consider myself an expert on the parallel features - I have a general understanding but that is all. Bill Long of Cray is the person I consider most knowledgeable about parallelism, though there are others on the committee who seem to understand it well.

I will ask Bill to take a look at this thread and see if he wants to offer an opinion.

**Comment by wclodius2** [comment]
@klausler your paper is written more as a feature request (this is the language of the standard, this is how I interpret the language, this is how it should be changed) when I think an interpretation request (this is the language of the standard, this is how I interpret the language, is my interpretation correct for that language, is my interpretation what was intended) would get more prompt attention.

**Comment by klausler** [comment]
> @klausler your paper is written more as a feature request (this is the language of the standard, this is how I interpret the language, this is how it should be changed) when I think an interpretation request (this is the language of the standard, this is how I interpret the language, is my interpretation correct for that language, is my interpretation what was intended) would get more prompt attention.

It's been 16 months, so prompt attention is a lost cause at this point.  As is J3.  I've removed myself from membership on that committee.

**Comment by sblionel** [comment]
I raised this question again yesterday on the J3 email list. You can read the thread [here](https://mailman.j3-fortran.org/pipermail/j3/2020-July/012229.html).  If I understand it correctly, these issues have been known for a while and are what prompted the addition in F2018 of the DEFAULT(NONE) locality specifier, as requiring the compiler to analyze the loop for possible sharing was difficult. The general opinion seems to be that explicitly specifying the locality of all variables is needed to enable parallelization, but that changing the F2008 behavior would break programs.

**Comment by klausler** [comment]
If anybody had tried to solve the problem of my specific example, they would have learned that the recently added locality clauses are not sufficient to its needs.  They accept only the names of whole variables (and, less important, can't distinguish between a pointer and its target).

**Comment by klausler** [comment]
DO CONCURRENT's locality rules are broken even apart from parallelization concerns.

Is the following a conforming program under Fortran 2018?

  PROGRAM ONE
    REAL :: A(2) = 0.
    INTEGER :: J
    DO CONCURRENT (J=1:2) SHARED(A)
      A(J) = A(J) + 1.
    END DO
    PRINT *, A
  END

Quoting 11.1.7.5(3):
  If a variable has SHARED locality, appearances of the variable within
  the DO CONCURRENT construct refer to the variable in the innermost
  executable construct or scoping unit that includes the DO CONCURRENT
  construct. If it is defined or becomes undefined during any iteration,
  it shall not be referenced, defined, or become undefined during any
  other iteration.

A has SHARED locality but is both defined and referenced in all iterations.

If A had unspecified locality (no SHARED(A) locality specifier), then
paragraph (4) (first bullet point) would apply:
  If a variable has unspecified locality, if it is referenced in an iteration
  it shall either be previously defined during that iteration, or shall not be
  defined or become undefined during any other iteration; if it is defined or
  becomes undefined by more than one iteration it becomes undefined when the
  loop terminates;

And the program would not be conforming, both for referencing A in each
iteration without a prior definition, and for using the undefinied
variable after the termination of the loop.

Can the word "variable" possibly refer to the elements of A, rather than
to its entirety?  Elsewhere in subclause 11.1.7, there are sites in
which the word "variable" clearly refers to whole arrays:

11.1.7.5(2), emphasis added
  A variable that has LOCAL or LOCAL_INIT locality is a construct entity with
  the same type, type parameters, and **rank** as the variable with the same
  name in the innermost executable construct or scoping unit that includes
  the DO CONCURRENT construct, and the outside variable is inaccessible by
  that name within the construct. ... If it is not a pointer, it has the
  same **bounds** as the outside variable.

11.1.7.5(4)
  If a variable has unspecified locality, ... if it is noncontiguous and
  is supplied as an actual argument corresponding to a contiguous
  INTENT (IN OUT) dummy argument in an iteration, it shall either be
  previously defined in that iteration or shall not be defined in any
  other iteration;

**Comment by sblionel** [comment]
Peter, would you please ask this on the J3 email list, of which you're still a member? I think you'd get a more authoritative response there. But to answer the question, "Can the word *variable* possibly refer to the elements of A", the answer is yes. R902 defines *variable* as including *designator* and *designator* (R901) includes *array-element*. 

I admit it can be a bit confusing where sometimes variable names are mentioned (and thus excluding array elements), but in the places where it just says *variable*, then an array element qualifies. In your example, the variable reference is `A(J)`, not `A`, and hence I would say that your example conforms, since no element of `A` is referenced or defined in more than one iteration.

**Comment by klausler** [comment]
Thanks for the reminder.

**Comment by certik** [comment]
I just asked on the J3 mailinglist to clarify the main problem raised in this issue:

https://mailman.j3-fortran.org/pipermail/j3/2020-July/012241.html

**Comment by certik** [comment]
@klausler here is an answer by Malcolm:

https://mailman.j3-fortran.org/pipermail/j3/2020-July/012244.html

If I understand it correctly, he says that to get maximum performance, one has to do `DEFAULT(SHARED)`. He provided some background and motivation why things are designed the way they are. He said "We discussed them at great length." I wish the discussion was archived somewhere, so that we don't need to repeat it. But at least we are discussing this now and it is archived now.

**Comment by klausler** [comment]
> @klausler here is an answer by Malcolm:
> 
> https://mailman.j3-fortran.org/pipermail/j3/2020-July/012244.html
> 
> If I understand it correctly, he says that to get maximum performance, one has to do `DEFAULT(SHARED)`. He provided some background and motivation why things are designed the way they are. He said "We discussed them at great length." I wish the discussion was archived somewhere, so that we don't need to repeat it. But at least we are discussing this now and it is archived now.

There is no `DEFAULT(SHARED)` in Fortran.  It might be a good idea, but it's not in the language.

**Comment by sblionel** [comment]
Right - to me, `DEFAULT(SHARED)` is as bad as `IMPLICIT`. Much better to explicitly specify the localities of all the variables in the block.

**Comment by klausler** [comment]
The value of a `DEFAULT(SHARED)`, if it actually were in the language, is that it would apply to variables that can't be named in an explicit locality specifier.

**Comment by certik** [comment]
@klausler I am still confused: with the current Fortran Standard and your example, compilers are required to put in (potentially) costly runtime checks, or is there a way to write it using explicit locality specifiers to parallelize efficiently?

**Comment by klausler** [comment]
> @klausler I am still confused: with the current Fortran Standard and your example, compilers are required to put in (potentially) costly runtime checks, or is there a way to write it using explicit locality specifiers to parallelize efficiently?

It depends.   If the loop is safe to parallelize in a straightforward manner -- any object that is modified is neither referenced nor modified in any other iteration -- and any potential cases that might be ambiguous due to the crazy default localization rule are confined to objects with simple names, then an explicit `SHARED` locality specifier would resolve the problem and permit parallelization.  That probably covers my original example, depending on the contents of the indirect indexing arrays.

For cases involving objects that don't have simple names, it's not clear.  (Take my original example, but make T a  component.)  This is why I said that the `DEFAULT(SHARED)` that Malcolm told you to use would be an improvement, if it existed in the language.

If there were a locality specifier (or other syntax) that changed the treatment of variables with unspecified locality so that the compiler would automatically localize only variables that were simple names of scalars, that would solve the problem.

**Comment by rouson** [comment]
@klausler  I'm teaching two Fortran training courses over the next 4 weeks and would greatly benefit from hearing any advice I could give attendees regarding `do concurrent`.  I was planning to recommend that every `do concurrent` contain a `default(none)`, but I see your comment above that this is insufficient because of limitations on what can appear in the locality specifiers.  Is there some non-trivial class of problems for which users can reasonably expect a compiler to be able to optimize `do concurrent`, whether by vectorization, parallelization, or offloading?   For example, I think you've mentioned in separate correspondence with me that the problem that arises in the code you provided in [comment 560517989] above would not occur in code that makes no use of indirect addressing or pointers.  I think there are many structured-grid scientific computing calculations that can be done without such features so it's not a severe restriction.  (By contrast, it's probably very hard to write unstructured-grid codes without indirect addressing or pointers.) Could a programmer reasonably expect that a compiler could effectively optimize  a `do concurrent` loop no indirect addressing, no pointers, and with a `default(none)` clause?  If not, is there any class of problems for which developers can reasonably expect a compiler to optimize `do concurrent` or are all bets off because of the issues discussed in this thread?

[comment 560517989]: https://github.com/j3-fortran/fortran_proposals/issues/62#issuecomment-560517989

**Comment by klausler** [comment]
I don't know of a Fortran compiler that recognizes the locality specifiers yet; the latest Intel compiler doesn't, anyway.  So even if they fixed the problems with default locality, your students would probably have to wait to use them.  This also applies to Malcolm Cohen's suggestion to use `DEFAULT(SHARED)`, which isn't even in the language.

I suggest that you treat `DO CONCURRENT` as a convenience for writing multidimensional iteration spaces with a single construct, no more.  I don't think that the implementations or standard for Fortran can help your parallelize sparse matrix codes without recourse to directive-based solutions.  For codes without pointers or indirection, one can try `DO CONCURRENT` and see if it works, but be ready to resort to directives on good old DO where it doesn't.  It's not performance portable.

**Comment by certik** [comment]
@klausler so coming to your original example, the following code would require runtime checks (thus not efficient), but it would work for any arrays K and L:
```fortran
SUBROUTINE FOO(N, A, B, T, K, L)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: N, K(N), L(N)
  REAL, INTENT(IN) :: A(N)
  REAL, INTENT(OUT) :: B(N)
  REAL, INTENT(INOUT) :: T(N)
  INTEGER :: J
  DO CONCURRENT (J=1:N)
    T(K(J)) = A(J)
    B(J) = T(L(J))
  END DO
END SUBROUTINE FOO
```

On the other hand, the following code would parallelize well, but it is now the responsibility of the user to ensure that K and L is not of the type `K=[1,1,2],L=[1,1,2]` (and similar):
```fortran
SUBROUTINE FOO(N, A, B, T, K, L)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: N, K(N), L(N)
  REAL, INTENT(IN) :: A(N)
  REAL, INTENT(OUT) :: B(N)
  REAL, INTENT(INOUT) :: T(N)
  INTEGER :: J
  DO CONCURRENT (J=1:N) DEFAULT(NONE) SHARED(T, A, B, K, L) LOCAL(J)
    T(K(J)) = A(J)
    B(J) = T(L(J))
  END DO
END SUBROUTINE FOO
```

Is the above analysis correct?

**Comment by klausler** [comment]
Surely you don't need to specifiy locality for K, L, or A -- they are not defined in the loop.  I don't read `DEFAULT(NONE)` to mean that the locality of variables that are merely referenced, not defined, has to be specified -- if so, it's crazy.  I'm also not sure whether it's allowed or necessary to specify the name of a loop index like `LOCAL(J)` -- that is at best redundant.

Declaring `SHARED(T)` makes a promise that T is not defined in any iteration and then referenced or defined in any other.  Whether that applies to the elements of T is not explicit in the standard, but it should be.  `T` is obviously defined and referenced in every iteration.  But if we use the obvious intent of the standard rather than the actual wording, and if it is true that no element of `T` (or `B`) that is defined in one iteration is referenced or defined in another, then the loop is trivially parallelizable.

**Comment by rouson** [comment]
@klausler Intel recently issued a beta release that is fully Fortran 2018 compliant so should support the locality specifiers.  For anyone interested in testing it, Intel's beta release includes two compilers:

1. `ifort`: Fortran 2018 compiler (the final release in this product line), which is in the oneAPI beta07.
2. `ifx`: LLVM-based Fortran 90 compiler (the beginning of a new product line). 

Below are the links you'll need to access the beta compilers.  There is no license needed.   You will need two toolkits:

1. First install "Intel(R) oneAPI Base Toolkit(Beta)":
https://software.intel.com/content/www/us/en/develop/tools/oneapi/base-toolkit.html

2.  Then install "Intel(R) oneAPI HPC Toolkit":
https://software.intel.com/content/www/us/en/develop/tools/oneapi/hpc-toolkit.html

**Comment by certik** [comment]
@klausler it seems there is confusion about the meaning of the Fortran Standard. But if we rewrite the code to use OpenMP:
```fortran
SUBROUTINE FOO(N, A, B, T, K, L)
  IMPLICIT NONE
  INTEGER, INTENT(IN) :: N, K(N), L(N)
  REAL, INTENT(IN) :: A(N)
  REAL, INTENT(OUT) :: B(N)
  REAL, INTENT(INOUT) :: T(N)
  INTEGER :: J
  !$OMP PARALLEL DEFAULT(NONE) SHARED(T, A, B, K, L, N) PRIVATE(J)
  !$OMP DO
  DO J = 1, N
    T(K(J)) = A(J)
    B(J) = T(L(J))
  END DO
  !$OMP END PARALLEL
END SUBROUTINE FOO
```
Then we can use, e.g., GFortran with the `-fopenmp` option to compile it. With GFortran, you can remove the `PRIVATE(J)`, but if you remove any of the `K`, `L` or `A`, you will get an error like:
```
a.f90:11:0:

     T(K(J)) = A(J)
 
Error: ‘a’ not specified in enclosing ‘parallel’
a.f90:8:0:

   !$OMP PARALLEL DEFAULT(NONE) SHARED(T, B, K, L, N)
 
Error: enclosing ‘parallel’
```
But this issue aside, in the above code, does the `SHARED(T)` apply to elements of `T`? If so, then the user is responsible to ensure that elements of `T` "are not defined in any iteration and then referenced or defined in any other", and the compiler will parallelize this efficiently?

**Comment by klausler** [comment]
> But this issue aside, in the above code, does the SHARED(T) apply to elements of T? If so, then the user is responsible to ensure that elements of T "are not defined in any iteration and then referenced or defined in any other", and the compiler will parallelize this efficiently?

Does `SHARED(T)` apply to the elements of `T`?  Clearly that's the intent, but I'm not sure whether the specification in the standard makes that clear.  What if `T` were a derived type with an array component -- would the locality specifier apply to the component?  What if `T` were a pointer -- would it apply to the pointer, or its target, or both?  It needs further work.

Is the user guaranteeing to the compiler that the stores to the elements are safe to parallelize?  Yes, that's the point.

Will the compiler parallelize it efficiently?  That's up to the implementation, of course.  But at least there shouldn't be any remaining need to generate code to resolve ambiguous "all equals" data dependences at runtime -- forwarding from stores to loads in the same iteration will either be via localized names or nonconflicting shared elements.

**Comment by certik** [comment]
Even the assignment to `B(J) = ...` assumes that `SHARED(B)` applies to elements of `B` (as guaranteed by the user), doesn't it? Otherwise the compiler would either have to figure out that the elements are independent (which in this case it can, but not in general), or insert runtime checks.

Regarding the derived type, I would assume as the user that the locality specifier applies to all components.

**Comment by klausler** [comment]
> Even the assignment to `B(J) = ...` assumes that `SHARED(B)` applies to elements of `B` (as guaranteed by the user), doesn't it? Otherwise the compiler would either have to figure out that the elements are independent (which in this case it can, but not in general), or insert runtime checks.

`B` is not referenced in the loop, and if (an element of) `B` is defined by more than iteration of the loop it may not be referenced afterwards the termination of the loop -- so there's really nothing for a compiler to do with it.

**Comment by rouson** [comment]
@sblionel Do you think there would be any support for designing a new feature that would be trivially distinct from `do concurrent` syntactically but would have semantics that compiler developers agree would be more amenable to optimization?  I'm imagining something like simply replacing `concurrent` with `independent` or some other word but keeping the syntax otherwise identical,  changing the default localization rules, and adding any other semantic restrictions that compiler developers agree would facilitate  optimizations such as vectorization, parallelization, or offloading are feasible.

**Comment by rouson** [comment]
@certik  I wonder if it would make sense to prototype my above suggestion in an open-source compiler or interpreter?

**Comment by certik** [comment]
Yes! I am sure Peter would be happy to help with Flang also. I have two summer students working on this very exact issue also for LFortran. Peter, Damian and Steve, would you be interested in a video call to get us on the same page?

In general this is one of the most important issues to get right in Fortran, and the main reason why people move away from Fortran around me.

**Comment by rouson** [comment]
@certik yes, I would be very interested in joining a call.  Scheduling it will be challenging: if it's not within the next 2 days, I probably won't be able to join until the week after next.  There's also a new gfortran developer who expressed interest in working on `do concurrent` so we should invite him too.

**Comment by klausler** [comment]
We're still figuring out how we're going to move ahead on this in f18 given J3's lack of a useful response.  Options range from command-line flags and inline directives to new syntax.  For example, f18 could assume better default locality rules by default and add a `-fstd-do-concurrent` flag to opt out.  Or add a `DO PARALLEL` with the fixed semantics and let the user use the preprocessor to port to it (`-Dconcurrent=parallel`) or away (`-Dparallel=concurrent`).  We'll keep you posted.  I don't think that a call would be terribly helpful at this point.

**Comment by rouson** [comment]
Hi Peter,

To quote a feature request that I once submitted to Intel, "Please, I'm
begging you to make standard behavior the default."  I can't count how many
hours of my life I've lost searching for a bug that turned out not to be a
bug but was instead a compiler making non-standard behavior the default
behavior.  It's really frustrating for anyone who uses multiple compilers
and expects the compiler to support the standard without the programmer
having to search for the proper flag to enforce the standard.  At an
absolute minimum, I'd hope there would be one all-encompassing flag like
Intel has "--standard-semantics," but even that only reduces the
frustration rather than eliminating it because one still has to remember to
add that to every project in which one uses a given compiler.

Damian

**Comment by sblionel** [comment]
@klausler , what exactly would be a "useful response" from J3? Three J3 members have explained in detail the situation and how the F18 locality specs essentially solve the problem, if programmers use them. It was also explained that the issues you raised had previously been discussed in detail. I'm failing to understand the continued antipathy here.

@rouson , I have said often that parallelism is not one of my strengths, so I rely on the advice and opinions of others who live this stuff. My (perhaps flawed) understanding is that as of F18, DO CONCURRENT enables more parallelism than F08 did. I know it is following what OpenMP did, and I have to believe that the folks developing OpenMP know what they are doing. I also have faith in J3, with representatives from multiple HPC compiler vendors, that DO CONCURRENT does what compiler vendors want for parallelism.

I have not yet seen any evidence that yet another parallel construct adds any value. The Fortran standard does not attempt to solve 100% of any possible programmer need. There is a balance between complexity and usability. Consider RANDOM_NUMBER, for an example of something I understand better. Does the spec for RANDOM_NUMBER satisfy every numerical analysis or statistical developer? Of course not, as some might want different distributions, etc., but it works well for a lot of cases. For those where it doesn't, there are libraries.

Regarding your asking Intel to make standard-conforming behavior the default - you're unlikely to get that from a long-history commercial compiler, as the howls from paying customers will be loud. As you note, Intel has a single flag that does what you want, and even offers [a way to set this system-wide (or user wide)](https://software.intel.com/content/www/us/en/develop/documentation/fortran-compiler-developer-guide-and-reference/top/compilation/using-configuration-files.html). Intel has changed defaults in places where it could be done transparently, such as automatic reallocation on assignment.

There's a related discussion going on in comp.lang.fortran where a couple of users are bashing the (few remaining) gfortran developers for having the temerity to have the compiler check by default for violation of standard rules that have been in place since F66. At least these are compile-time messages for violations of the standard - think of what the response would be if programs that were standard-conforming before (because the standard didn't specify something) behaved differently at run-time without any obvious cause.

The Fortran standard has always been in the position of asking the programmer to follow certain rules, and if they do, then they are rewarded with expected behavior. The rules for DO CONCURRENT about definition changes across iterations are there to allow compilers to parallelize, and there exist compilers that can and do parallelize these when the rules are followed. I don't pretend to claim that every possible loop body can be parallelized, but my understanding is that the current syntax and rules are good enough for most usages. If one wants even more control, then look at OpenMP. I can't find much enthusiasm for examples of constructs that have always been difficult to parallelize, even with the most capable of methods.

**Comment by klausler** [comment]
The response that came closest to a solution involved syntax (`DEFAULT(SHARED)`) that strengthens the stipulations from the program(mer) about the lack of any need for the default localization rules in order to safely execute iterations in parallel.  But `DEFAULT(SHARED)` doesn't exist in the language, and would necessitate wordy and error-prone `LOCAL()` specifiers for non-shared variables (or `BLOCK`).  A `DO CONCURRENT` plus `DEFAULT(SHARED)` with further automatic localization for the obvious scalar variable names that are unconditionally defined in each iteration would solve the problem.  `DEFAULT(SHARED_EXCEPT_NO_BRAINER_SCALARS)` seems like too much typing, though.

The discussion on comp.lang.fortran about extant code that associates scalar actual arguments with dummy arrays across implicit interfaces exposes the distinction between _de facto_ and _de jure_ standards.   IMO, something that's worked for decades with mulitple compilers is not something that should be messed with absent very good reasons.  Nobody really benefits when something universally used like NetCDF stops compiling.

**Comment by rouson** [comment]
@klausler if the problems you raised could be addressed by adding a new `default(...)` clause, then that seems like a lot of bang for the buck.   My understanding is that it was once common for features to start as compiler extensions before entering the standard.   I would definitely give such an extension a try if it appears in a compiler.  It seems like something that would be very easy to incorporate via preprocessor macros when working with a compiler that supports it. I would also back a proposal for such a feature at the appropriate time (presumably 202Y).

**Comment by klausler** [comment]
https://reviews.llvm.org/D86556

We’re still trying to figure out what we’re going to do with DO CONCURRENT in f18 (flang). I wrote a short documentation file to describe its basic problem: a programmer writing parallel code with DO CONCURRENT may not enjoy parallel execution because the standard imposes weaker requirements on the body of DO CONCURRENT that suffice only to allow safe serial execution of iterations in arbitrary order, and a compiler can assume only that those weaker requirements have been respected and thus has no general way to know that concurrency is safe – even when the programmer knows that it is.

Please let me know where the note can be made more clear, correct or complete. I’m trying to summarize the problems in one place in the source base to which we can direct new users in the future when they need to know why DO CONCURRENT isn’t working as one might expect. If you have LLVM reviewing permissions, you can comment in place if you like. Thanks!

**Comment by rouson** [comment]
Hi Peter,

Thanks for writing this.  I'll copy the list of solutions under
consideration, (re)number them, and intersperse comments and questions.

1. Add syntax (e.g., `DO PARALLEL` or `DO CONCURRENT() DEFAULT(PARALLEL)`)
   by which we can inform the compiler that it should localize only
   the obvious cases of simple local scalars.
   Such syntax seems unlikely to ever be standardized, so its usage
   would be nonportable.

This would be my first choice, especially given the historical precedents
for features being supported first as extensions before becoming part of
the standard.  Coarray features might be the most appropriate precedent
because they relate to parallelism.  Didn't Cray support coarrays for years
before they became standard?  I imagine it was a much heavier lift to get
all the coarray features into the standard than either of the above
choices. In the interim, a preprocessor could easily switch either of the
above to standard syntax.

2. Add a command-line option &/or a source directive to stipulate
   the stronger guarantees.  Obvious non-parallelizable usage in the
construct
   would elicit a stern warning.  The `DO CONCURRENT` loops in the source
   would continue to be portable to other compilers.

The command-line option would be my third choice, but it's really not such
a bad choice.  The code would remain standard and portable and anyone
serious about optimizing code is likely to spend time investigating the
command-line options for each compiler they use.  I'd rather not clutter
the code with compiler-specific directives, especially if other compilers
go down a similar path but choose different syntax.  If the directive
approach is taken, I hope the directive is a potential candidate for
inclusion into the OpenMP standard, but that seems a lot less likely than
getting option 1 into the Fortran standard.

3. Assume that these stronger conditions hold by default, and add a
command-line
   option &/or a source directive to "opt out" back to the standard
   behavior in the event that the program contains one of those
   non-parallelizable `DO CONCURRENT` loops that should never have
   been possible to write in a conforming program in the first place.
   Programmers writing parallel `DO CONCURRENT` loops would get what
   they wanted without ever knowing that there's a problem with the
   standard.
   But this could lead to non-standard behavior for codes that depend,
   accidentally or not, on non-parallelizable implicit localization.

As I mentioned previously, I can't count the number of hours of my life
that I've lost looking for a bug in my code only to find eventually that
there was no bug.  I had just forgotten to pass a flag to get the compiler
to comply with the standard.  I beg you to make standard behavior the
default behavior.

4. Accept the standard as it exists, do the best job of automatic
   parallelization that can be done, and refer dissatisfied users to J3.
   This would be avoiding the problem.

This would be my second choice.  I think of this as being embedded in the
first option.  Create a new syntax that does what you want and accept the
standard as is when users choose the standard syntax.

I hope this is helpfu.

Damian

**Comment by certik** [comment]
In my opinion none of these options are ideal.

I think the ideal solution is to bring this back to the Fortran Standard Committee and have serious discussion about what the problem is and how to fix that, and then fix it.

**Comment by klausler** [comment]
Damian: Thanks for your response.

Are you using DO CONCURRENT to write parallel code?

**Comment by rouson** [comment]
Hi Peter,

I use DO CONCURRENT for what it communicates to the reader and for the
discipline it imposes on the programmer.  I would love it if that then also
leads to performance improvements, whether they be by parallelization or
vectorization or offloading.  Short of that, I would also be satisfied if
OpenMP can be used to parallelize or vectorize or offload DO CONCURRENT in
the absence of the compiler doing it automatically.  I just submitted a
proposal to do work for a government agency wherein performance
improvements associated with DO CONCURRENT could play a very useful role --
again whether by compiler automation or via OpenMP.

Damian

**Comment by klausler** [comment]
What do you think that "it [DO CONCURRENT] communicates to the reader"?

**Comment by klausler** [comment]
> In my opinion none of these options are ideal.
> 
> I think the ideal solution is to bring this back to the Fortran Standard Committee and have serious discussion about what the problem is and how to fix that, and then fix it.

Oh, sure; all of these options are competing for "second best", under the assumption that J3 doesn't fix DO CONCURRENT.

**Comment by rouson** [comment]
It communicates that all of the constraints in the standard apply.  As
examples, it communicates that every procedure invoked is PURE and there
are no segment-ordering statements.

As a general rule, if one program statement covers a subset of the cases
that another program statement covers, I choose the more specific one in
hopes that doing so decreases the time it takes for the reader to
understand the code.  I think of DO CONCURRENT as covering a subset of the
loops that can be written with just DO so I start with DO CONCURRENT but
resort to DO when I have a compelling reason to do so.

Similarly, I think of SELECT CASE as covering a subset of the conditional
branching that can be accomplished with IF THEN... ELSE IF... ELSE... END
IF.  Choosing the more specific statement communicates information that
might otherwise require the reader to read through each branch to discover
the logic of the branching.

Damian

P.S.  On a related note, I just completed an exercise in co-developing a
mini-app in which we attempted to take functional programming to a logical
extreme: with the exception of I/O procedures, every procedure in the
mini-app is a PURE function with no local variables -- just associations
wherein a name is associated with a function result or expression and is
therefore immutable state.  I was surprised at just how far this got us and
with the ultimate clarity of the code.  At just of shy of 1,500 lines, it
was the longest code I've ever written that worked almost flawlessly on
first testing.  This experience emboldens me to write PURE functions more
and more and one benefit will be that all of my functions can be invoked
inside DO CONCURRENT.  I suspect that code written in this style will be
much easier for the programmer to parallelize even in the absence of
automatic compiler parallelization.  One of the first places I encountered
the notion that functional programming aids parallelization was in a 1991
Lawrence Livermore technical report entitled "Retire Fortran? A debate
rekindled."  It's fascinating to me just how much of what that paper was
calling for in a new language (SISAL) instead resembles what Fortran
eventually became.

**Comment by klausler** [comment]
Thanks, that helps.  I agree that programs should use `PURE` wherever they can do so.  But in the context of `DO CONCURRENT`, the purity requirement is not an end in itself, but a means to another end, namely concurrency.  The language shouldn't define `DO CONCURRENT` in such a way as to exclude a long list of concurrency-breaking behavior without making the list complete -- that punishes the programmer who *is* writing concurrent code but is still not rewarded with concurrency because a compiler can't assume that following the rules in the standard is sufficient.

**Comment by certik** [comment]
@klausler thanks for your efforts and I agree with you this is the second best.

As we get more experience as a community with using `do concurrent` in codes, I think Damian, myself and others will be able to bring any potential limitations to the committee to get further discussed and fixed.

**Comment by everythingfunctional** [comment]
After seeing all the discussions around this, I am now under the following impression. The standard chose to use `DO CONCURRENT` but only specified constraints for (and if we had a time machine should have used) `DO UNORDERED`. It seems many (if not most) people assumed that the constraints actually were sufficient to allow for parallelism, but it would seem to be highly unlikely the committee would "break existing code" by adding constraints to `DO CONCURRENT`. Thus it seems we will need to introduce (and I would be in favor of) a new keyword to allow a way of specifying constraints sufficient for parallelism, most likely `DO PARALLEL`.

Of course, for those of us who considered concurrent and parallel to be synonyms it makes the language a bit more confusing, but is probably the only truly backwards compatible way to get the necessary constraints added. It seems to me quite similar to the way `PURE` procedures don't have sufficient constraints, hence the proposal to introduce `SIMPLE` procedures.

**Comment by klausler** [comment]
> After seeing all the discussions around this, I am now under the following impression. The standard chose to use `DO CONCURRENT` but only specified constraints for (and if we had a time machine should have used) `DO UNORDERED`. It seems many (if not most) people assumed that the constraints actually were sufficient to allow for parallelism, but it would seem to be highly unlikely the committee would "break existing code" by adding constraints to `DO CONCURRENT`. Thus it seems we will need to introduce (and I would be in favor of) a new keyword to allow a way of specifying constraints sufficient for parallelism, most likely `DO PARALLEL`.
> 
> Of course, for those of us who considered concurrent and parallel to be synonyms it makes the language a bit more confusing, but is probably the only truly backwards compatible way to get the necessary constraints added. It seems to me quite similar to the way `PURE` procedures don't have sufficient constraints, hence the proposal to introduce `SIMPLE` procedures.

See subclause 4.3.3 in Fortran 2018 for some precedents in which features from earlier standards are disallowed, including some features from `DO CONCURRENT`.  Backward compatibility to mistakes is fixable.

**Comment by klausler** [comment]
I've described the outstanding problems with `DO CONCURRENT` in the [LLVM documenation](https://github.com/llvm/llvm-project/blob/master/flang/docs/DoConcurrent.md).  Still not sure what we're going to do in the flang compilers -- there's good arguments for both standard conformance as well as for just doing the right thing.

**Comment by FortranFan** [comment]
@klausler wrote Nov. 17, 2020 7:01 PM EST:

> .. Still not sure what we're going to do in the flang compilers -- there's good arguments for both standard conformance as well as for just doing the right thing.

c.f. https://mailman.j3-fortran.org/pipermail/j3/2020-July/012244.html where J3 "effectively" suggested something which is not yet in the standard.

So, is it possible for flang compilers to do both!?

That is, first have a standard-conforming implementation of DO CONCURRENT.

But then also consider an "Experimental" edition of flang compiler(s) that attempts to do the "right thing", perhaps via a "DEFAULT(SHARED)" or some suitable extension that is Fortrannic and which can then be proposed for Fortran 202Y as further improvement to be incorporated into the standard?

**Comment by klausler** [comment]
> @klausler wrote Nov. 17, 2020 7:01 PM EST:
> 
> > .. Still not sure what we're going to do in the flang compilers -- there's good arguments for both standard conformance as well as for just doing the right thing.
> 
> c.f. https://mailman.j3-fortran.org/pipermail/j3/2020-July/012244.html where J3 "effectively" suggested something which is not yet in the standard.
> 
> So, is it possible for flang compilers to do both!?
> 
> That is, first have a standard-conforming implementation of DO CONCURRENT.
> 
> But then also consider an "Experimental" edition of flang compiler(s) that attempts to do the "right thing", perhaps via a "DEFAULT(SHARED)" or some suitable extension that is Fortrannic and which can then be proposed for Fortran 202Y as further improvement to be incorporated into the standard?

Anything is possible, but these are all just second-best alternatives to J3 just fixing the problems.

**Comment by wyphan** [comment]
Hi @klausler , now that my Google Summer of Code project proposal for `DO CONCURRENT` support in GFortran has been accepted, I'd like to set up a meeting to discuss about this. (I wish GitHub has a messaging feature...)

If you happen to be at the Fortran Discourse forums, please send me a PM with your NVIDIA email address so I can send you a calendar invite to the meeting I'm in the process of setting up with Jeff Larkin, Güray Özen, and the folks at Predictive Science who published arXiv:2110.10151 about the `DO CONCURRENT` implementation in NVIDIA `nvfortran`. Otherwise you can reach me at `wileam [at] phan [dot] codes`.

**Comment by klausler** [comment]
I have reviewed the [document](https://github.com/llvm/llvm-project/blob/main/flang/docs/DoConcurrent.md) for LLVM Fortran that describes the problems with `DO CONCURRENT` in the standard language, and unfortunately nothing has changed on the language standards front since I wrote it in 2020.  In particular, nothing has been done to fix `DO CONCURRENT` in the draft Fortran 202X standards (apart maybe from `SIMPLE` procedures), and there has been no further discussion of the problems, at least not that I can find.  Perhaps somebody on WG5/J3 may take it up for Fortran 202Y but that is of course a long ways out.

**Comment by certik** [comment]
>  Perhaps somebody on WG5/J3 may take it up for Fortran 202Y but that is of course a long ways out.

@klausler you have the most knowledge on this particular issue (since you wrote the Flang document!). Do you think you could please submit proposals for 2Y to fix this? I'll help champion it and advocate for it, but it would take me much longer to write up than it would take for you, since you have thought about all the details here and what needs to be done.

**Comment by klausler** [comment]
My [2019 paper](https://j3-fortran.org/doc/year/19/19-134.txt), which was ignored by J3, remains my favored solution; recycle it if you like.  It would be nice if the semantics of locality specifiers with regard to pointers and `ASSOCIATE/SELECT TYPE` names would also be clarified, too, as being invalid for `LOCAL()`.

**Comment by rouson** [comment]
@klausler could you share with us what course of action NVIDIA chose for offloading `do concurrent`?  If, for example, the compiler allows me pass a flag that toggles offloading off and on at compile time, I would be happy.  I would always include `default(none)`, which at least would force me to think through locality so that I have a good understanding of whether it's safe to offload.

**Comment by rouson** [comment]
> The three problems with `DO CONCURRENT (...) DEFAULT(NONE)` are that (1) you have to name every data item in every loop in a locality clause, (2) the F'2018 locality clauses only allow simple names and (3) the semantics of pointers in locality clauses are not defined.

This is helpful.  I hope the NVIDIA representative(s) on J3 can champion your suggestions.  I'd like to see these issues addressed but I would be a poor champion because I have little personal use for most of the features that break `do concurrent`.  It seems that the problematic features mostly involve pointers (which I rarely use), indirect addressing (which I seldom use),  or module variables (which I mostly avoid) so I wouldn't have many useful insights in any of the discussions that arise around a new proposal.

> 
> A better (but not best) suggestion was to use `DEFAULT(SHARED)` but the problem with that is that it doesn't exist.

A J3 member in the aforementioned mailing list discussion that `default(shared)` was considered by the committee and that it was decided that `default(none)` would be less error-prone.  A reference was made to the complexities of OpenMP's `default(shared)`.  I don't know enough to have an opinion.  I just want to say that the suggestion was considered and another path taken for reasons that were provided.

> 
> So I still recommend that the default implicit localization rule be changed (perhaps by syntax) to pertain only to names that could appear in an explicit `LOCAL()` clause. The fact that it was defined so that it applies to array elements and components is accidental; if fixed, it would work the way everybody assumes that it already does.

Maybe but I think the committee gives more weight to what the committee intended than to what users might incorrectly assume.  I think the unstated goal is to make the standard clear, consistent, and useful.  If it's unclear, add a note.  If it's inconsistent, fix it.   If it's clear and consistent but not useful, _replace_ the feature much like `do concurrent` effectively replaces `forall`, but don't break existing code that relies upon the feature as long as that code has an unambiguous interpretation that is consistent with what those who wrote the standard intended.  This is just my observation of how the committee operates and I'm ok with it. It's why I usually recommend that people attend several meetings before making a proposal.  It can take some time to read the room and understand the dynamics and the implicit aims.

**Comment by wyphan** [comment]
@klausler This is exactly why I want to start discussion. I plan to push for `DEFAULT` as a GNU extension and implement it accordingly during the GSoC. I'm inviting you, @jefflarkin, and @grypp from NVIDIA, Ron Caplan and Miko Stulajter from @predsci because they published [arXiv:2110.10151](https://arxiv.org/abs/2110.10151) as the first real-world use case for `DO CONCURRENT`, as well as my GSoC mentors Tobias Burnus and @tschwinge from GCC/Siemens. And as I indicated before, other interested folks in this thread are welcome too. Tentatively it will be held between June 6-10, which is the week after ISC '22. Hopefully, upon successful implementation as a GNU extension, the standards committee will consider to include it in Fortran 202Y.

**Comment by klausler** [comment]
Another WG5/J3 meeting has come and gone with no recorded action on fixing DO CONCURRENT.  The HPC subgroup didn't even submit a report on their assigned F'202Y discussion items.

At this point, the best implementation option appears to me to be to ignore the broken standard and assume that the default localization rules apply only to variables that could have appeared in an explicit LOCAL clause.  J3 has had three years to fix this and done nothing.

**Comment by sblionel** [comment]
Your item is on the list of things being considered, and there was quite a bit of discussion, but no action at this time. There is quite a bit of disagreement on the matter, especially with some of the claims, but it is being taken seriously. You're correct that the HPC subgroup didn't submit a paper with initial comments, but there will be further discussion.

It;'s too bad that you chose to withdraw from the committee since you obviously have a passion for the issues.

**Comment by klausler** [comment]
> Your item is on the list of things being considered, and there was quite a bit of discussion, but no action at this time. There is quite a bit of disagreement on the matter, especially with some of the claims, but it is being taken seriously. You're correct that the HPC subgroup didn't submit a paper with initial comments, but there will be further discussion.
> 
> It;'s too bad that you chose to withdraw from the committee since you obviously have a passion for the issues.

I became convinced that J3's current process is incapable of producing quality work.  The best that I can do is describe the bugs in the standard as I encounter them as an implementer, so that they're documented and you can fix them or not in the standard as you choose.  It's not that different of a situation from being a user of a buggy compiler -- one works around the bugs, but still reports them responsibly in the hope that something might be done about them before they affect other users.

**Comment by klausler** [comment]
The plan for HPC features in Fortran 202Y (https://j3-fortran.org/doc/year/23/23-146.txt) omits any mention of fixing `DO CONCURRENT`.

**Comment by jeffhammond** [comment]
I took another look at this, particularly Peter's examples.

The following is https://github.com/j3-fortran/fortran_proposals/issues/62#issuecomment-560517989 except with the read from T removed.

```fortran
subroutine foo(N, A, B, T, K, L)
  implicit none
  integer, intent(in) :: N, K(N), L(N)
  real, intent(in) :: A(N)
  real, intent(out) :: B(N)
  real, intent(inout) :: T(N)
  integer :: J
  do concurrent (J=1:N)
    ! During execution, K(J) is always 1.
    T(K(J)) = A(J)
  end do
end subroutine foo
```

As no locality is specified, we can refer to the following:

> If a variable has unspecified locality, if it is referenced in an iteration it shall either be previously defined during that iteration, or shall not be defined or become undefined during any other iteration; if it is defined or becomes undefined by more than one iteration it becomes undefined when the loop terminates;

Writing to `T` in multiple iterations violates this and causes `T` to become undefined.  Unless all of the elements of `K` are unique, this program contains a data race.

If we add `shared(T)` to the code, nothing changes.

> If it is defined or becomes undefined during any iteration, it shall not be referenced, defined, or become undefined during any other iteration.

Note that I am interpreting "variable" to mean the element of an array, not the whole array, even though it's unclear, because if I interpret "variable" as the whole array, it is impossible to use `DO CONCURRENT` with arrays.

Can someone tell me what is wrong with my thinking?

@AnastasiaStulova

**Comment by klausler** [comment]
You deleted the reference to `T(K(J))`, but the interpretation of that reference is the whole point of the text that you quote.

From Fortran's perspective there are no such things as "data races" in DO CONCURRENT.  It's not a parallel programming construct.  `T(1)` is undefined after the loop ends (if there's more than one iteration) because of the definitions in multiple iterations, but within each iteration, it is well defined in the current standard.  This is the problem in a nutshell.

**Comment by jeffhammond** [comment]
My point is that, there is a race on T in both your program and mine, and unless WG5 believes that race conditions are legal and defined, the interpretation of your program does not matter, because it has undefined behavior before it gets to the interesting part.

**Comment by klausler** [comment]
It is meaningless to talk about race conditions in serial code.  DO CONCURRENT, despite its name, is defined as a serial construct.  RYOS.  F'202X 11.1.7.4.3 paragraph 3: "The block of a DO CONCURRENT construct is executed for every active combination of the index-name values. Each execution of the block is an iteration. The executions may occur in any order."

**Comment by jeffhammond** [comment]
it was intended to allow parallel implementations.  i am proceeding with the intent to make parallelism a reasonable implementation.  given that Fujitsu, Cray, Intel and NVIDIA all implement DC with parallelism in a wide range of cases, i believe that allowing data races was the mistake, not parallelism.

**Comment by klausler** [comment]
Yes, that is entirely my point.  DO CONCURRENT's default locality rules were badly defined and allow non-parallelizable data accesses to be written in conforming code.

**Comment by jeffhammond** [comment]
default locality rules are not even the issue.  as i posted on the WG5 list earlier today, the following program is problematic, despite having explicit locality and none of the interesting properties of your example.

```fortran
program main
  integer :: k
  real :: x
  x = 0.0
  do concurrent (k=1:2) shared(x)
    x = real(k)
  end do
  print*,x
end program main
```

**Comment by klausler** [comment]
Your example is clearly non-conforming, and should remain so.  And it's not relevant to this particular issue.
